glibc: update requirements

maxim-belkin opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/linuxbrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/linuxbrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?
- [ ] Have you included the output of `brew gist-logs <formula>` of the build failure if your PR fixes a build failure. Please quote the exact error message.

-----
Updating some requirements based on notes in `INSTALL` file in `glibc` source code.

- Update Gawk requirement (minimum required version 3.1.2)
- Add Make requirement (minimum required version 3.79)
- Add Sed requirement (any version)
- Update comments:
   * binutils: minimum required version 2.22
   * Linux kernel headers 2.6.32

CC @sjackman

sjackman commented:
I'd prefer not to add complexity to this already tangly formula to address issues that have not yet arisen in the real world. It'd be appropriate if it were addressing a user reported issue.

sjackman commented on Formula/glibc.rb:
```diff
@@ -18,9 +18,15 @@ def message
   satisfy(:build_env => false) do
     # Returning which("gawk") causes a cyclic dependency.
-    !which("gawk").nil?
+    !which("gawk").nil? and gawk_version >= MINIMUM_GAWK_VERSION
```
```suggestion
    which("gawk").present?
```

sjackman commented on Formula/glibc.rb:
```diff
@@ -64,9 +108,11 @@ class Glibc < Formula
   end
 
-  depends_on "binutils" => :build # binutils 2.20 or later is required
+  depends_on "binutils" => :build # binutils 2.22 or later is required
```
```suggestion
  depends_on "binutils" => :build
```
This comment was stale, and will likely become stale again. Better to remove it.

sjackman commented on Formula/glibc.rb:
```diff
@@ -64,9 +108,11 @@ class Glibc < Formula
+  depends_on "binutils" => :build # binutils 2.22 or later is required
   depends_on GawkRequirement => :build
-  depends_on "linux-headers" => :build # Linux kernel headers 2.6.19 or later are required
+  depends_on "linux-headers" => :build # Linux kernel headers 2.6.32 or later are required
```
```suggestion
  depends_on "linux-headers" => :build
```
This comment was stale, and will likely become stale again. Better to remove it.

maxim-belkin commented:
> I'd prefer not to add complexity to this already tangly formula to address issues that have not yet arisen in the real world. The existing checks addressed issues experienced and reported by users.

I'm not sure what complexity you mean as the changes are easy to understand and are designed to explain users which tool on their system is not up-to-date. I expect these requirements to become more stringent over time so having this code in place should be useful when we move to the next version of glibc (btw, when is it going to happen?)

sjackman commented on Formula/glibc.rb:
```diff
@@ -54,6 +60,44 @@ def message
+class MakeRequirement < Requirement
+  fatal true
+
+  MINIMUM_MAKE_VERSION = "3.79".freeze
```
GNU Make 3.79 was released in the year 2000. I do not believe it is worth the code to verify that the version is sufficient.

sjackman commented on Formula/glibc.rb:
```diff
@@ -54,6 +60,44 @@ def message
+  end
+
+  satisfy(:build_env => false) do
+    make_version >= MINIMUM_MAKE_VERSION
```
```suggestion
    which("make").present?
```

sjackman commented on Formula/glibc.rb:
```diff
@@ -54,6 +60,44 @@ def message
+  fatal true
+
+  satisfy(:build_env => false) do
+    !which("sed").nil?
```
```suggestion
    which("sed").present?
```

sjackman commented on Formula/glibc.rb:
```diff
@@ -54,6 +60,44 @@ def message
+      Install make with your host package manager if you have sudo access.
+        sudo (apt-get|yum|dnf) install make
+    EOS
+  end
```
Please factor out `GawkRequirement.message` into its own function, and use that in each of these message functions.

maxim-belkin replied:
do you want only apt-get and yum? or
apt-get
yum
dnf
zypper
... ?

sjackman replied:
`apt-get` and `yum` are sufficient.

maxim-belkin replied:
Done!

iMichka commented:
I think we need to use “safe_popen_read” everywhere instead of “popen_read”.

iMichka commented:
Aren’t sed / make / gawk part of the build essentials, which is required by our documentation?

maxim-belkin commented:
> Aren’t sed / make / gawk part of the build essentials, which is required by our documentation?

`sed` is Essential
`make` is Build Essential
`gawk` is neither

`Gawk` causes some troubles, because `which` is executed with PATH set to
```
"/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/shims/scm:/usr/bin:/bin:/usr/sbin:/sbin"
```
and it does not find gawk... 🤔

sjackman commented on Formula/glibc.rb:
```diff
@@ -18,9 +18,15 @@ def message
   satisfy(:build_env => false) do
     # Returning which("gawk") causes a cyclic dependency.
-    !which("gawk").nil?
+    which("gawk").present? and gawk_version >= MINIMUM_GAWK_VERSION
```
```suggestion
    which("gawk").present?
```
Gawk 3.1.2 was released in 2003. I do not believe it is worth adding the code to verify that the version is sufficient.
https://ftp.gnu.org/gnu/gawk/

maxim-belkin replied:
what cutoff do you use to decide whether it should or shouldn't verify software version?

sjackman replied:
The oldest version of `glibc` that we support is 2.13, and it was release on 2011-02-01, so let's say ~2010.
https://docs.brew.sh/Homebrew-on-Linux#requirements
https://sourceware.org/glibc/wiki/Glibc%20Timeline

maxim-belkin commented:
@sjackman, I've updated the PR (sorry for the long wait -- it fell off my radar)

maxim-belkin commented:
This is not ready yet

maxim-belkin commented:
I think this is ready now.

maxim-belkin commented:
Question about `gawk`: shall we mention `brew install gawk`?

```sh
$ sudo apt-get remove sed
$ brew install -s glibc
glibc: GNU Awk 3.1.2 or later is required to build glibc.
Install gawk with your host package manager if you have sudo access.
  sudo apt-get install gawk
  sudo yum install gawk
glibc: Sed is required to build glibc.
Install sed with your host package manager if you have sudo access.
  sudo apt-get install sed
  sudo yum install sed
Error: Unsatisfied requirements failed this build.
```

sjackman commented on Formula/glibc.rb:
```diff
@@ -20,15 +20,19 @@ class GawkRequirement < Requirement
+    self.class.message_template(tool: "gawk", nice_name: "GNU Awk", min_version: "3.1.2")
+  end
+
+  def self.message_template(tool: "", nice_name: "", min_version: "")
```
Can `message_template` be a bare function outside of any class?

maxim-belkin replied:
🤷🏻‍♂️￼

maxim-belkin replied:
I can do something like
```rb
def Requirement::message_template(...)
...
end
```
and then use `Requirement::message_template` instead of `GawkRequirement.message_template`. Not sure if there is a better solution.

sjackman commented:
> Question about gawk: shall we mention brew install gawk?

No, because installing the `gawk` bottle requires first installing `glibc`, which the user is currently trying to install.

sjackman commented:
Squash merge without updating the bottles.

maxim-belkin commented:
@sjackman, Could you please review this once again?

@iMichka, Could you please merge this once this is approved?

sjackman commented on Formula/glibc.rb:
```diff
@@ -1,34 +1,50 @@
-    Glibc.version >= OS::Linux::Glibc.system_version
-  end
-
+class GlibcRequirement < Requirement
```
Minor point, but I'd suggest renaming `GlibcRequirement` to `GlibcBaseRequirement`.

sjackman commented on Formula/glibc.rb:
```diff
@@ -1,34 +1,50 @@
-class GawkRequirement < Requirement
+class GawkRequirement < GlibcRequirement
+  fatal true
+  satisfy(build_env: false) { which(TOOL).present? }
```
Can these `fatal` and `satisfy` lines be moved to the base class?

sjackman replied:
Bump?

maxim-belkin replied:
Unfortunately, thees can't be moved to GlibcBaseRequirement -- at least, I don't know of a way to do that preserving the functionality. Perhaps, Mike knows how to do that...

sjackman replied:
It's fine as is if it's not easy.

sjackman commented on Formula/glibc.rb:
```diff
@@ -1,34 +1,50 @@
     <<~EOS
-      Your system's glibc version is #{OS::Linux::Glibc.system_version}, and Homebrew's glibc version is #{Glibc.version}.
-      Installing a version of glibc that is older than your system's can break formulae installed from source.
+      #{[tool, version].join(" ").strip} is required to build glibc.
```
```suggestion
      #{[tool, version].compact.join(" ")} is required to build glibc.
```

sjackman commented on Formula/glibc.rb:
```diff
@@ -1,34 +1,50 @@
+  fatal true
+  satisfy(build_env: false) { which(TOOL).present? }
+  TOOL = "sed".freeze
+  VERSION = "".freeze
```
```suggestion
  VERSION = nil
```

sjackman commented on Formula/glibc.rb:
```diff
@@ -1,34 +1,50 @@
+  VERSION = "".freeze
+end
+
+class BrewedGlibcNotOlderRequirement < Requirement
```
Please leave this class at the top. It'll make the diff tidier as a bonus.

maxim-belkin commented:
@iMichka, this PR is ready 🎉

maxim-belkin commented:
I think the answer is:
- Yes, if this image is used to build `glibc
- No, if some other image is used.

I think you build glibc using another image, so homebrew/brew doesn't have to have `gawk` (it has `awk`)

sjackman commented:
As far as I know, there is no problem building `glibc` with the normal `homebrew/ubuntu16.04:master` Docker image, except that it doesn't have `gawk`. Debian does include `mawk` by default, symlinked to `awk`, but that is not sufficient to build `glibc`.
```console
$ brew install -s glibc
…
configure: error: 
*** These critical programs are missing or too old: gawk
*** Check the INSTALL file for required versions.
```

sjackman commented:
> CI is now skipped. Should we add gawk to the build image?

It's not needed until we decide to rebuild `glibc`, which isn't until next year, so let's punt on this issue for now and leave the image as is. We can merge this PR as is with a squash merge.

iMichka commented:
Let's do this then. Thanks to both of you for the update.
